File: SDKs/java/src/main/java/com/intelliversex/sdk/core/IVXManager.java
package com.intelliversex.sdk.core;

import com.heroiclabs.nakama.Client;
import com.heroiclabs.nakama.DefaultClient;
import com.heroiclabs.nakama.Session;
import com.heroiclabs.nakama.api.Account;
import com.heroiclabs.nakama.api.LeaderboardRecord;
import com.heroiclabs.nakama.api.LeaderboardRecordList;
import com.heroiclabs.nakama.api.Rpc;
import com.heroiclabs.nakama.api.StorageObjectAcks;
import com.heroiclabs.nakama.api.StorageObjects;
import com.heroiclabs.nakama.api.WriteStorageObject;
import com.heroiclabs.nakama.api.ReadStorageObjectId;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.protobuf.StringValue;

import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;
import java.util.prefs.BackingStoreException;
import java.util.prefs.Preferences;

/**
 * Central manager for the IntelliVerseX Java/Android SDK.
 * <p>
 * Wraps the Nakama Java client with auth, profile, wallet, leaderboards,
 * storage, and RPC.
 * <p>
 * <b>Thread-safety:</b> All public methods that mutate state use synchronized
 * blocks on this instance. Callbacks and futures are safe to use from any thread.
 * <p>
 * <b>Android note:</b> {@code java.util.prefs.Preferences} may not work on all
 * Android devices. For Android projects, call
 * {@link #setPreferences(Preferences)} with a custom {@code Preferences}
 * implementation backed by {@code SharedPreferences} <i>before</i> calling
 * {@link #initialize(IVXConfig)}.
 */
public class IVXManager {
    public static final String SDK_VERSION = "5.1.0";

    private static final String PREF_SESSION_TOKEN = "ivx_session_token";
    private static final String PREF_REFRESH_TOKEN = "ivx_refresh_token";
    private static final String PREF_DEVICE_ID = "ivx_device_id";

    private static volatile IVXManager instance;

    private IVXConfig config;
    private Client client;
    private volatile Session session;
    private volatile boolean initialized = false;
    private final Gson gson = new Gson();
    private Preferences prefs = Preferences.userNodeForPackage(IVXManager.class);

    private final Map<String, List<Consumer<Object>>> listeners = new ConcurrentHashMap<>();

    private IVXManager() {}

    public static IVXManager getInstance() {
        if (instance == null) {
            synchronized (IVXManager.class) {
                if (instance == null) {
                    instance = new IVXManager();
                }
            }
        }
        return instance;
    }

    /**
     * Replace the default {@link Preferences} store. On Android, provide one backed
     * by SharedPreferences. Must be called <i>before</i> {@link #initialize(IVXConfig)}.
     */
    public synchronized void setPreferences(Preferences prefs) {
        Objects.requireNonNull(prefs, "prefs must not be null");
        this.prefs = prefs;
    }

    public boolean isInitialized() { return initialized; }
    public Client getClient() { return client; }
    public Session getSession() { return session; }
    public String getUserId() { return session != null ? session.getUserId() : ""; }
    public String getUsername() { return session != null ? session.getUsername() : ""; }
    public boolean hasValidSession() { return session != null && !session.isExpired(); }

    // ─── Events ─────────────────────────────────────────────────

    @SuppressWarnings("unchecked")
    public <T> void on(String event, Consumer<T> handler) {
        listeners.computeIfAbsent(event, k -> new CopyOnWriteArrayList<>()).add((Consumer<Object>) handler);
    }

    @SuppressWarnings("unchecked")
    private <T> void emit(String event, T data) {
        List<Consumer<Object>> handlers = listeners.get(event);
        if (handlers != null) {
            for (Consumer<Object> h : handlers) {
                h.accept(data);
            }
        }
    }

    // ─── Init ───────────────────────────────────────────────────

    public synchronized void initialize(IVXConfig config) {
        Objects.requireNonNull(config, "config must not be null");
        this.config = config;

        this.client = new DefaultClient(
                config.getNakamaServerKey(),
                config.getNakamaHost(),
                config.getNakamaPort(),
                config.isUseSSL()
        );

        this.initialized = true;
        log("SDK initialized — " + config.getBaseUrl());
        emit("initialized", null);
    }

    // ─── Auth (blocking) ────────────────────────────────────────

    public void authenticateDevice(String deviceId) {
        ensureInitialized();
        String resolvedId = (deviceId == null || deviceId.isEmpty()) ? getPersistentDeviceId() : deviceId;

        try {
            Session newSession = client.authenticateDevice(resolvedId).get();
            onAuthSuccess(newSession);
        } catch (Exception e) {
            log("Auth failed: " + e.getMessage());
            emit("authError", e.getMessage());
        }
    }

    public void authenticateEmail(String email, String password, boolean create) {
        ensureInitialized();

        try {
            Session newSession = client.authenticateEmail(email, password, create).get();
            onAuthSuccess(newSession);
        } catch (Exception e) {
            emit("authError", e.getMessage());
        }
    }

    public void authenticateGoogle(String token) {
        ensureInitialized();

        try {
            Session newSession = client.authenticateGoogle(token).get();
            onAuthSuccess(newSession);
        } catch (Exception e) {
            emit("authError", e.getMessage());
        }
    }

    public void authenticateApple(String token) {
        ensureInitialized();

        try {
            Session newSession = client.authenticateApple(token).get();
            onAuthSuccess(newSession);
        } catch (Exception e) {
            emit("authError", e.getMessage());
        }
    }

    public void authenticateCustom(String customId) {
        ensureInitialized();

        try {
            Session newSession = client.authenticateCustom(customId).get();
            onAuthSuccess(newSession);
        } catch (Exception e) {
            emit("authError", e.getMessage());
        }
    }

    // ─── Auth (async) ───────────────────────────────────────────

    /**
     * Authenticates with a device ID asynchronously.
     *
     * @param deviceId device identifier, or {@code null}/empty to auto-generate
     * @return a {@link CompletableFuture} that completes with the user ID on success
     */
    public CompletableFuture<String> authenticateDeviceAsync(String deviceId) {
        ensureInitialized();
        String resolvedId = (deviceId == null || deviceId.isEmpty()) ? getPersistentDeviceId() : deviceId;

        return CompletableFuture.supplyAsync(() -> {
            try {
                Session newSession = client.authenticateDevice(resolvedId).get();
                onAuthSuccess(newSession);
                return newSession.getUserId();
            } catch (Exception e) {
                log("Async auth failed: " + e.getMessage());
                emit("authError", e.getMessage());
                throw new RuntimeException("Authentication failed", e);
            }
        });
    }

    /**
     * Authenticates with email/password asynchronously.
     */
    public CompletableFuture<String> authenticateEmailAsync(String email, String password, boolean create) {
        ensureInitialized();

        return CompletableFuture.supplyAsync(() -> {
            try {
                Session newSession = client.authenticateEmail(email, password, create).get();
                onAuthSuccess(newSession);
                return newSession.getUserId();
            } catch (Exception e) {
                emit("authError", e.getMessage());
                throw new RuntimeException("Authentication failed", e);
            }
        });
    }

    /**
     * Authenticates with a Google token asynchronously.
     */
    public CompletableFuture<String> authenticateGoogleAsync(String token) {
        ensureInitialized();

        return CompletableFuture.supplyAsync(() -> {
            try {
                Session newSession = client.authenticateGoogle(token).get();
                onAuthSuccess(newSession);
                return newSession.getUserId();
            } catch (Exception e) {
                emit("authError", e.getMessage());
                throw new RuntimeException("Authentication failed", e);
            }
        });
    }

    /**
     * Authenticates with an Apple token asynchronously.
     */
    public CompletableFuture<String> authenticateAppleAsync(String token) {
        ensureInitialized();

        return CompletableFuture.supplyAsync(() -> {
            try {
                Session newSession = client.authenticateApple(token).get();
                onAuthSuccess(newSession);
                return newSession.getUserId();
            } catch (Exception e) {
                emit("authError", e.getMessage());
                throw new RuntimeException("Authentication failed", e);
            }
        });
    }

    /**
     * Authenticates with a custom ID asynchronously.
     */
    public CompletableFuture<String> authenticateCustomAsync(String customId) {
        ensureInitialized();

        return CompletableFuture.supplyAsync(() -> {
            try {
                Session newSession = client.authenticateCustom(customId).get();
                onAuthSuccess(newSession);
                return newSession.getUserId();
            } catch (Exception e) {
                emit("authError", e.getMessage());
                throw new RuntimeException("Authentication failed", e);
            }
        });
    }

    // ─── Session ────────────────────────────────────────────────

    public synchronized boolean restoreSession() {
        String token = prefs.get(PREF_SESSION_TOKEN, "");
        String refresh = prefs.get(PREF_REFRESH_TOKEN, "");

        if (token.isEmpty()) return false;

        session = DefaultClient.restoreSession(token, refresh);
        if (session.isExpired()) {
            session = null;
            return false;
        }

        log("Session restored for user: " + session.getUserId());
        syncMetadata();
        return true;
    }

    public synchronized void clearSession() {
        session = null;
        prefs.put(PREF_SESSION_TOKEN, "");
        prefs.put(PREF_REFRESH_TOKEN, "");
        flushPrefs();
        log("Session cleared");
    }

    // ─── Profile ────────────────────────────────────────────────

    /**
     * Fetches the current player's profile (blocking).
     *
     * @return an {@link IVXProfile} or {@code null} on failure
     */
    public IVXProfile fetchProfile() {
        ensureSession();

        try {
            Account account = client.getAccount(session).get();
            IVXProfile profile = new IVXProfile(
                    account.getUser().getId(),
                    account.getUser().getUsername(),
                    account.getUser().getDisplayName(),
                    account.getUser().getAvatarUrl(),
                    account.getUser().getLangTag(),
                    account.getUser().getMetadata(),
                    account.getWallet()
            );
            emit("profileLoaded", profile);
            return profile;
        } catch (Exception e) {
            emit("error", e.getMessage());
            return null;
        }
    }

    /**
     * Fetches the current player's profile asynchronously.
     */
    public CompletableFuture<IVXProfile> fetchProfileAsync() {
        ensureSession();

        return CompletableFuture.supplyAsync(() -> fetchProfile());
    }

    public void updateProfile(String displayName, String avatarUrl, String langTag) {
        ensureSession();

        try {
            client.updateAccount(session, null, displayName, avatarUrl, langTag, null).get();
            log("Profile updated");
        } catch (Exception e) {
            emit("error", e.getMessage());
        }
    }

    // ─── Wallet ─────────────────────────────────────────────────

    public String fetchWallet() {
        return callRpc("hiro_economy_list", "{}");
    }

    public String grantCurrency(String currencyId, long amount) {
        JsonObject currencies = new JsonObject();
        currencies.addProperty(currencyId, amount);
        JsonObject payload = new JsonObject();
        payload.add("currencies", currencies);
        return callRpc("hiro_economy_grant", gson.toJson(payload));
    }

    // ─── Leaderboard ────────────────────────────────────────────

    public void submitScore(String leaderboardId, long score) {
        ensureSession();

        try {
            client.writeLeaderboardRecord(session, leaderboardId, score).get();
            log("Score submitted: " + score + " to " + leaderboardId);
        } catch (Exception e) {
            emit("error", e.getMessage());
        }
    }

    public List<Map<String, Object>> fetchLeaderboard(String leaderboardId, int limit) {
        ensureSession();

        try {
            LeaderboardRecordList result = client.listLeaderboardRecords(session, leaderboardId, null, null, limit).get();
            List<Map<String, Object>> records = new ArrayList<>();
            for (LeaderboardRecord r : result.getRecordsList()) {
                Map<String, Object> record = new HashMap<>();
                record.put("ownerId", r.getOwnerId());

                StringValue usernameVal = r.getUsername();
                record.put("username", (usernameVal != null && !usernameVal.getValue().isEmpty())
                        ? usernameVal.getValue() : "");

                record.put("score", r.getScore());
                record.put("rank", r.getRank());
                records.add(record);
            }
            return records;
        } catch (Exception e) {
            emit("error", e.getMessage());
            return Collections.emptyList();
        }
    }

    // ─── Storage ────────────────────────────────────────────────

    public void writeStorage(String collection, String key, String valueJson) {
        ensureSession();

        try {
            WriteStorageObject obj = WriteStorageObject.newBuilder()
                    .setCollection(collection)
                    .setKey(key)
                    .setValue(valueJson)
                    .setPermissionRead(1)
                    .setPermissionWrite(1)
                    .build();
            client.writeStorageObjects(session, obj).get();
            log("Storage write: " + collection + "/" + key);
        } catch (Exception e) {
            emit("error", e.getMessage());
        }
    }

    public String readStorage(String collection, String key) {
        ensureSession();

        try {
            ReadStorageObjectId id = ReadStorageObjectId.newBuilder()
                    .setCollection(collection)
                    .setKey(key)
                    .setUserId(getUserId())
                    .build();
            StorageObjects result = client.readStorageObjects(session, id).get();
            if (result.getObjectsCount() > 0) {
                return result.getObjects(0).getValue();
            }
            return "{}";
        } catch (Exception e) {
            emit("error", e.getMessage());
            return "{}";
        }
    }

    // ─── RPC ────────────────────────────────────────────────────

    /**
     * Calls a server RPC endpoint (blocking).
     *
     * @param rpcId       the RPC identifier
     * @param payloadJson JSON payload string; {@code null} is normalized to "{}"
     * @return the response payload JSON, or "{}" on failure
     */
    public String callRpc(String rpcId, String payloadJson) {
        ensureSession();
        String safePayload = (payloadJson == null) ? "{}" : payloadJson;

        try {
            Rpc result = client.rpc(session, rpcId, safePayload).get();
            log("RPC " + rpcId + " response received");
            return result.getPayload();
        } catch (Exception e) {
            log("RPC " + rpcId + " failed: " + e.getMessage());
            emit("error", e.getMessage());
            return "{}";
        }
    }

    /**
     * Calls a server RPC endpoint asynchronously.
     */
    public CompletableFuture<String> callRpcAsync(String rpcId, String payloadJson) {
        ensureSession();
        String safePayload = (payloadJson == null) ? "{}" : payloadJson;

        return CompletableFuture.supplyAsync(() -> {
            try {
                Rpc result = client.rpc(session, rpcId, safePayload).get();
                log("RPC " + rpcId + " response received");
                return result.getPayload();
            } catch (Exception e) {
                log("RPC " + rpcId + " failed: " + e.getMessage());
                emit("error", e.getMessage());
                throw new RuntimeException("RPC " + rpcId + " failed", e);
            }
        });
    }

    // ─── Internal ───────────────────────────────────────────────

    private synchronized void onAuthSuccess(Session newSession) {
        this.session = newSession;
        prefs.put(PREF_SESSION_TOKEN, newSession.getAuthToken());
        prefs.put(PREF_REFRESH_TOKEN, newSession.getRefreshToken());
        flushPrefs();
        log("Authenticated — UserId: " + newSession.getUserId());
        syncMetadata();
        emit("authSuccess", newSession.getUserId());
    }

    private void syncMetadata() {
        if (!hasValidSession()) return;
        JsonObject meta = new JsonObject();
        meta.addProperty("sdk_version", SDK_VERSION);
        meta.addProperty("platform", System.getProperty("os.name", "unknown"));
        meta.addProperty("engine", "java");
        meta.addProperty("java_version", System.getProperty("java.version", "unknown"));

        JsonObject payload = new JsonObject();
        payload.add("metadata", meta);

        try {
            callRpc("ivx_sync_metadata", gson.toJson(payload));
        } catch (Exception ignored) {
            // Metadata sync failure is non-fatal
        }
    }

    private synchronized String getPersistentDeviceId() {
        String id = prefs.get(PREF_DEVICE_ID, "");
        if (id.isEmpty()) {
            id = UUID.randomUUID().toString();
            prefs.put(PREF_DEVICE_ID, id);
            flushPrefs();
        }
        return id;
    }

    private void flushPrefs() {
        try {
            prefs.flush();
        } catch (BackingStoreException e) {
            log("WARNING: Could not flush preferences — " + e.getMessage());
        }
    }

    private void ensureInitialized() {
        if (!initialized || client == null) {
            throw new IllegalStateException("IntelliVerseX SDK not initialized. Call initialize() first.");
        }
    }

    private void ensureSession() {
        ensureInitialized();
        if (!hasValidSession()) {
            throw new IllegalStateException("No valid Nakama session. Authenticate first.");
        }
    }

    private void log(String message) {
        if (config != null && config.isEnableDebugLogs()) {
            System.out.println("[IntelliVerseX] " + message);
        }
    }
}

File: SDKs/flutter/lib/src/ivx_manager.dart
import 'dart:convert';

import 'package:nakama/nakama.dart';

import 'ivx_config.dart';
import 'types.dart';

typedef IVXEventHandler = void Function(dynamic data);

/// Central coordinator for the IntelliVerseX Flutter SDK.
///
/// Provides singleton access to authentication, profile management,
/// wallet/economy, leaderboards, cloud storage, RPC, and real-time sockets.
class IVXManager {
  static IVXManager? _instance;

  Client? _client;
  Session? _session;
  IVXConfig _config = const IVXConfig();
  bool _initialized = false;
  final Map<IVXEvent, Set<IVXEventHandler>> _listeners = {};
  String? _cachedDeviceId;

  IVXManager._();

  static IVXManager get instance => _instance ??= IVXManager._();

  /// Reset the singleton (useful for testing).
  static void resetInstance() => _instance = null;

  Client? get client => _client;
  Session? get session => _session;
  bool get isInitialized => _initialized;
  String get userId => _session?.userId ?? '';
  String get username => _session?.username ?? '';

  bool get hasValidSession {
    if (_session == null) return false;
    final now = DateTime.now();
    return _session!.hasExpired(now) == false;
  }

  // ---------------------------------------------------------------------------
  // Initialization
  // ---------------------------------------------------------------------------

  void initialize(IVXConfig config) {
    config.validate();
    _config = config;

    _client = getNakamaClient(
      host: _config.nakamaHost,
      httpPort: _config.nakamaPort,
      serverKey: _config.nakamaServerKey,
      ssl: _config.useSSL,
    );

    _initialized = true;
    _log('SDK v$sdkVersion initialized - '
        '${_config.useSSL ? "https" : "http"}://${_config.nakamaHost}:${_config.nakamaPort}');
    _emit(IVXEvent.initialized);
  }

  // ---------------------------------------------------------------------------
  // Events
  // ---------------------------------------------------------------------------

  void on(IVXEvent event, IVXEventHandler handler) {
    _listeners.putIfAbsent(event, () => {});
    _listeners[event]!.add(handler);
  }

  void off(IVXEvent event, IVXEventHandler handler) {
    _listeners[event]?.remove(handler);
  }

  void _emit(IVXEvent event, [dynamic data]) {
    final handlers = _listeners[event];
    if (handlers == null) return;
    for (final handler in handlers) {
      handler(data);
    }
  }

  // ---------------------------------------------------------------------------
  // Authentication
  // ---------------------------------------------------------------------------

  Future<void> authenticateDevice([String? deviceId]) async {
    _ensureInitialized();
    final resolvedId = deviceId ?? _generateDeviceId();
    try {
      _session = await _client!.authenticateDevice(deviceId: resolvedId);
      _onAuthSuccess();
    } catch (e) {
      final error = _toIVXError(e);
      _emit(IVXEvent.authError, error);
      throw error;
    }
  }

  Future<void> authenticateEmail(
    String email,
    String password, {
    bool create = false,
  }) async {
    _ensureInitialized();
    try {
      _session = await _client!.authenticateEmail(
        email: email,
        password: password,
        create: create,
      );
      _onAuthSuccess();
    } catch (e) {
      final error = _toIVXError(e);
      _emit(IVXEvent.authError, error);
      throw error;
    }
  }

  Future<void> authenticateGoogle(String token) async {
    _ensureInitialized();
    try {
      _session = await _client!.authenticateGoogle(token: token);
      _onAuthSuccess();
    } catch (e) {
      final error = _toIVXError(e);
      _emit(IVXEvent.authError, error);
      throw error;
    }
  }

  Future<void> authenticateApple(String token) async {
    _ensureInitialized();
    try {
      _session = await _client!.authenticateApple(token: token);
      _onAuthSuccess();
    } catch (e) {
      final error = _toIVXError(e);
      _emit(IVXEvent.authError, error);
      throw error;
    }
  }

  Future<void> authenticateCustom(String customId) async {
    _ensureInitialized();
    try {
      _session = await _client!.authenticateCustom(id: customId);
      _onAuthSuccess();
    } catch (e) {
      final error = _toIVXError(e);
      _emit(IVXEvent.authError, error);
      throw error;
    }
  }

  void clearSession() {
    _session = null;
    _log('Session cleared');
  }

  // ---------------------------------------------------------------------------
  // Profile
  // ---------------------------------------------------------------------------

  Future<IVXProfile> fetchProfile() async {
    _ensureSession();
    try {
      final account = await _client!.getAccount(_session!);
      final profile = IVXProfile(
        userId: account.user?.id ?? '',
        username: account.user?.username ?? '',
        displayName: account.user?.displayName ?? '',
        avatarUrl: account.user?.avatarUrl ?? '',
        langTag: account.user?.langTag ?? '',
        metadata: _safeDecodeJson(account.user?.metadata),
        wallet: _safeDecodeWallet(account.wallet),
      );
      _emit(IVXEvent.profileLoaded, profile);
      return profile;
    } catch (e) {
      final error = _toIVXError(e);
      _emit(IVXEvent.error, error);
      throw error;
    }
  }

  Future<void> updateProfile({
    String? displayName,
    String? avatarUrl,
    String? langTag,
  }) async {
    _ensureSession();
    try {
      await _client!.updateAccount(
        _session!,
        displayName: displayName,
        avatarUrl: avatarUrl,
        langTag: langTag,
      );
      _log('Profile updated');
    } catch (e) {
      final error = _toIVXError(e);
      _emit(IVXEvent.error, error);
      throw error;
    }
  }

  // ---------------------------------------------------------------------------
  // Wallet / Economy
  // ---------------------------------------------------------------------------

  Future<Map<String, dynamic>> fetchWallet() async {
    final result = await callRpc('hiro_economy_list');
    _emit(IVXEvent.walletUpdated, result);
    return result;
  }

  Future<Map<String, dynamic>> grantCurrency(
      String currencyId, int amount) async {
    return callRpc('hiro_economy_grant',
        jsonEncode({'currencies': {currencyId: amount}}));
  }

  // ---------------------------------------------------------------------------
  // Leaderboards
  // ---------------------------------------------------------------------------

  Future<void> submitScore(String leaderboardId, int score) async {
    _ensureSession();
    try {
      await _client!.writeLeaderboardRecord(
        _session!,
        leaderboardId: leaderboardId,
        score: score,
      );
      _log('Score submitted: $score to $leaderboardId');
    } catch (e) {
      final error = _toIVXError(e);
      _emit(IVXEvent.error, error);
      throw error;
    }
  }

  Future<List<IVXLeaderboardRecord>> fetchLeaderboard(
    String leaderboardId, {
    int limit = 20,
  }) async {
    _ensureSession();
    try {
      final result = await _client!.listLeaderboardRecords(
        _session!,
        leaderboardId: leaderboardId,
        limit: limit,
      );
      final records = (result.records ?? [])
          .map((r) => IVXLeaderboardRecord(
                ownerId: r.ownerId ?? '',
                username: r.username?.value ?? '',
                score: int.tryParse(r.score ?? '0') ?? 0,
                rank: int.tryParse(r.rank ?? '0') ?? 0,
              ))
          .toList();
      _emit(IVXEvent.leaderboardFetched, records);
      return records;
    } catch (e) {
      final error = _toIVXError(e);
      _emit(IVXEvent.error, error);
      throw error;
    }
  }

  // ---------------------------------------------------------------------------
  // Storage
  // ---------------------------------------------------------------------------

  Future<void> writeStorage(
      String collection, String key, Map<String, dynamic> value) async {
    _ensureSession();
    try {
      await _client!.writeStorageObjects(
        _session!,
        objects: [
          StorageObjectWrite(
            collection: collection,
            key: key,
            value: jsonEncode(value),
            permissionRead: 1,
            permissionWrite: 1,
          ),
        ],
      );
      _log('Storage write: $collection/$key');
    } catch (e) {
      final error = _toIVXError(e);
      _emit(IVXEvent.error, error);
      throw error;
    }
  }

  Future<Map<String, dynamic>?> readStorage(
      String collection, String key) async {
    _ensureSession();
    try {
      final result = await _client!.readStorageObjects(
        _session!,
        objectIds: [
          StorageObjectId(
            collection: collection,
            key: key,
            userId: userId,
          ),
        ],
      );
      if (result.objects != null && result.objects!.isNotEmpty) {
        final data = _safeDecodeJson(result.objects!.first.value);
        _emit(IVXEvent.storageRead, data);
        return data;
      }
      _emit(IVXEvent.storageRead, null);
      return null;
    } catch (e) {
      final error = _toIVXError(e);
      _emit(IVXEvent.error, error);
      throw error;
    }
  }

  // ---------------------------------------------------------------------------
  // RPC
  // ---------------------------------------------------------------------------

  Future<Map<String, dynamic>> callRpc(String rpcId,
      [String payload = '{}']) async {
    _ensureSession();
    try {
      final result = await _client!.rpc(_session!, id: rpcId, payload: payload);
      _log('RPC $rpcId response received');
      final data = _safeDecodeJson(result.payload);
      _emit(IVXEvent.rpcResponse, {'rpcId': rpcId, 'data': data});
      return data;
    } catch (e) {
      final error = _toIVXError(e);
      _log('RPC $rpcId failed: ${error.message}');
      _emit(IVXEvent.error, error);
      throw error;
    }
  }

  // ---------------------------------------------------------------------------
  // Internal helpers
  // ---------------------------------------------------------------------------

  void _onAuthSuccess() {
    _log('Authenticated - UserId: ${_session!.userId}');
    _syncMetadata();
    _emit(IVXEvent.authSuccess, _session!.userId);
  }

  Future<void> _syncMetadata() async {
    if (!hasValidSession) return;
    try {
      await callRpc(
        'ivx_sync_metadata',
        jsonEncode({
          'metadata': {
            'sdk_version': sdkVersion,
            'platform': 'dart',
            'engine': 'flutter',
          }
        }),
      );
    } catch (_) {
      // Non-fatal: metadata sync failure should not break the app
    }
  }

  String _generateDeviceId() {
    if (_cachedDeviceId != null) return _cachedDeviceId!;
    final now = DateTime.now().microsecondsSinceEpoch.toRadixString(36);
    final rand1 = (DateTime.now().millisecond * 31337).toRadixString(36);
    final rand2 = now.hashCode.abs().toRadixString(36);
    _cachedDeviceId = 'dart-$now-$rand1-$rand2';
    return _cachedDeviceId!;
  }

  Map<String, dynamic> _safeDecodeJson(dynamic value) {
    if (value is Map<String, dynamic>) return value;
    if (value is String && value.isNotEmpty) {
      try {
        final decoded = jsonDecode(value);
        if (decoded is Map<String, dynamic>) return decoded;
      } catch (_) {}
    }
    return {};
  }

  Map<String, num> _safeDecodeWallet(dynamic value) {
    final raw = _safeDecodeJson(value);
    return raw.map((k, v) => MapEntry(k, (v is num) ? v : 0));
  }

  IVXError _toIVXError(dynamic e) {
    if (e is IVXError) return e;
    return IVXError(code: -1, message: e.toString());
  }

  void _ensureInitialized() {
    if (!_initialized || _client == null) {
      final err = const IVXError(
        code: -1,
        message: 'SDK not initialized. Call initialize() first.',
      );
      _emit(IVXEvent.error, err);
      throw err;
    }
  }

  void _ensureSession() {
    _ensureInitialized();
    if (!hasValidSession) {
      final err = const IVXError(
        code: -1,
        message: 'No valid session. Authenticate first.',
      );
      _emit(IVXEvent.error, err);
      throw err;
    }
  }

  void _log(String message) {
    if (_config.enableDebugLogs) {
      print('[IntelliVerseX] $message');
    }
  }
}
